### PR TITLE
CAZ-1871 Support for multiple VRNs and different rates per day in POST /payments

### PR DESCRIPTION
**Pull request by radekmorytko**

### CODE REVIEW

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java [hunk 1]
```diff
@@ -1,44 +1,136 @@
...
+import uk.gov.caz.psr.controller.exception.InvalidRequestPayloadException;
 
 @Value
 @Builder
+@Slf4j
 public class InitiatePaymentRequest {
 
+  private static final Map<Function<InitiatePaymentRequest, Boolean>, String> validators =
+      ImmutableMap.<Function<InitiatePaymentRequest, Boolean>, String>builder()
+          .put(request -> Objects.nonNull(request.getCleanAirZoneId()),
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: Any chance to make it more eyes friendly?
For example (code written here without verification):
```
.put(cleanAirZoneIdNotNull(), "message"),
.put(returnUrlNotEmpty(), "message"),
.put(vrnInAllTransactionsNotEmpty(),  "message"),
```

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java [hunk 2]
```diff
@@ -1,44 +1,136 @@
...
+import org.springframework.util.StringUtils;
+import uk.gov.caz.psr.controller.exception.InvalidRequestPayloadException;
 
 @Value
 @Builder
+@Slf4j
 public class InitiatePaymentRequest {
 
+  private static final Map<Function<InitiatePaymentRequest, Boolean>, String> validators =
+      ImmutableMap.<Function<InitiatePaymentRequest, Boolean>, String>builder()
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: We should definitely use [MapPreservingOrderBuilder](https://github.com/InformedSolutions/JAQU-CAZ-Accounts-API/blob/develop/src/main/java/uk/gov/caz/accounts/util/MapPreservingOrderBuilder.java) here, for example like [here](https://github.com/InformedSolutions/JAQU-CAZ-Accounts-API/blob/0e5116493abc586dd2c70cda4c41dd2c8622cb7b/src/main/java/uk/gov/caz/accounts/dto/AccountCreationRequestDto.java#L41)

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java [hunk 3]
```diff
@@ -1,44 +1,136 @@
...
+          .put(request -> !CollectionUtils.isEmpty(request.getTransactions()),
+              "'transactions' cannot be null or empty")
+          .put(request -> allTransactionsMatch(request,
+              transaction -> StringUtils.hasText(transaction.getVrn())),
+              "'vrn' cannot be null or empty")
+          .put(request -> allTransactionsMatch(request,
+              transaction -> transaction.getVrn().length() >= 1
+                  && transaction.getVrn().length() <= 15), "'vrn' length must be from 1 to 15")
+          .put(request -> allTransactionsMatch(request,
+              transaction -> Objects.nonNull(transaction.getCharge())), "'charge' cannot be null")
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: Do we want (or will want?) to have more precise error messages saying that attribute of `Transaction` with such and such `vrn` is invalid? Probably not?

But maybe I would rephrase error message like this: `'charge' in all transactions cannot be null`

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java [hunk 4]
```diff
@@ -1,44 +1,136 @@
...
+  @ApiModelProperty(value = "${swagger.model.descriptions.payments-initiate.transactions}")
+  List<Transaction> transactions;
 
-  @ApiModelProperty(value = "${swagger.model.descriptions.payments-initiate.amount}")
-  @NotNull
-  @Positive
-  Integer amount;
+  @Value
+  @Builder(toBuilder = true)
+  public static class Transaction {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: Javadoc

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java [hunk 5]
```diff
@@ -1,44 +1,136 @@
...
+        throw new InvalidRequestPayloadException(message);
+      }
+    });
+  }
+
+  /**
+   * Verifies if {@code request} does not contain duplicated vehicle entrants.
+   */
+  private static boolean containsNoDuplicatedEntrants(InitiatePaymentRequest request) {
+    Map<String, List<LocalDate>> transactionDatesByVrn = groupByVrn(request);
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: A lot is going on here and it is hard to see what it is. Please try to extract to smaller helpers.

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java (hunk 2 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: I have double checked the documentation and ImmutableMaps preserve insertion order, so we should be fine using it:
> By default, a Builder will generate maps that iterate over entries in the order they were inserted into the builder, equivalently to LinkedHashMap

source: https://guava.dev/releases/28.0-jre/api/docs/com/google/common/collect/ImmutableMap.Builder.html

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java (hunk 1 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: Good idea, changed

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java (hunk 3 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: I changed the validation error messages.

File: src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java (hunk 5 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/dto/InitiatePaymentRequest.java]: I tried to make it more readable.
